api/http/handler/endpointedge/async.go

```diff
@@ -0,0 +1,312 @@
+}
+
+// TODO: yup, next step is for these to be JSONDiff's and to be rehydrated
+func (handler *Handler) saveSnapshot(endpoint *portainer.Endpoint, payload AsyncRequest) time.Time {
+	for stackID, status := range payload.StackStatus {
+		stack, err := handler.DataStore.EdgeStack().EdgeStack(portainer.EdgeStackID(stackID))
+		// TODO: work out what we can do with the errors
+		if err == nil {
+			stack.Status[*status.EndpointID] = portainer.EdgeStackStatus{
+				Type:       *status.Status,
+				Error:      status.Error,
+				EndpointID: *status.EndpointID,
+			}
+			err = handler.DataStore.EdgeStack().UpdateEdgeStack(stack.ID, stack)
+		}
+	}
+
+	switch endpoint.Type {
+	// case portainer.AzureEnvironment:
+	// 	return time.Now()
+	// case portainer.KubernetesLocalEnvironment, portainer.AgentOnKubernetesEnvironment, portainer.EdgeAgentOnKubernetesEnvironment:
+	// 	var ksnap portainer.KubernetesSnapshot
+	// 	err := request.DecodeAndValidateJSONPayload(AsyncRequest.Snapshot, &ksnap)
+	// 	if err != nil {
+	// 		// an "" request ~~ same as {}
+	// 		logrus.WithError(err).WithField("payload", r).Debug("decode payload")
+	// 	}
+	// 	endpoint.Kubernetes.Snapshots = []portainer.KubernetesSnapshot{ksnap}
+
+	// 	return time.Unix(ksnap.Time, 0)
+	case portainer.DockerEnvironment, portainer.AgentOnDockerEnvironment, portainer.EdgeAgentOnDockerEnvironment:
+		logrus.Debug("Got a Docker Snapshot")
+		dsnap, valid := payload.Snapshot.(portainer.DockerSnapshot)
+		if !valid {
+			// an "" request ~~ same as {}
+			logrus.Error("snapshot isn't valid")
+			return time.Time{}
+		}
+		endpoint.Snapshots = []portainer.DockerSnapshot{dsnap}
+		return time.Unix(dsnap.Time, 0)
+	}
+
+	return time.Time{}
+}
+
+func (handler *Handler) sendCommandsSince(endpoint *portainer.Endpoint, commandId string) []JSONPatch {
+	var commandList []JSONPatch
+
+	// TODO: get a list of all stacks for this endpoint - endpointStatusInspect
+	// TODO: start with being a rollup of all the endpointEdgeStackInspect for an endpoint
+	// TODO: later, figure out if it is scalable to do diff's, as it means the server needs to store what it sent to all million agents (if the database had time based versioning, this would be trivial...)
+	logrus.WithField("endpoint", endpoint.Name).WithField("from command", commandId).Debug("Sending commands")
+
+	// schedules := []edgeJobResponse{}
+	tunnel := handler.ReverseTunnelService.GetTunnelDetails(endpoint.ID)
+	for _, job := range tunnel.Jobs {
+		schedule := edgeJobResponse{
+			ID:             job.ID,
+			CronExpression: job.CronExpression,
+			CollectLogs:    job.Endpoints[endpoint.ID].CollectLogs,
+			Version:        job.Version,
+		}
+
+		file, err := handler.FileService.GetFileContent("/", job.ScriptPath)
+		if err != nil {
+			// TODO: this should maybe just skip thi job?
+			logrus.WithError(err).Error("Unable to retrieve Edge job script file")
+			continue
+		}
+
+		schedule.Script = base64.RawStdEncoding.EncodeToString(file)
```

Comment:
seems kind of weird to have to compute this every time a requests is received, maybe parse the script file only when needed?